crates/matrix-sdk-base/src/client.rs

```diff
@@ -1186,23 +1206,24 @@ impl BaseClient {
             new_rooms.invite.insert(room_id, new_info);
         }
 
-        // TODO remove this, we're processing account data events here again
+        // We're processing direct state events here separately
         // because we want to have the push rules in place before we process
         // rooms and their events, but we want to create the rooms before we
         // process the `m.direct` account data event.
-        let has_new_direct_room_data = response.account_data.events.iter().any(|raw_event| {
-            raw_event
-                .deserialize()
-                .map(|event| event.event_type() == GlobalAccountDataEventType::Direct)
-                .unwrap_or_default()
-        });
+        let has_new_direct_room_data = global_account_data_events
+            .iter()
+            .any(|event| event.event_type() == GlobalAccountDataEventType::Direct);
         if has_new_direct_room_data {
-            self.handle_account_data(&response.account_data.events, &mut changes).await;
+            self.process_direct_rooms(global_account_data_events, &mut changes).await;
         } else if let Ok(Some(direct_account_data)) =
             self.store.get_account_data_event(GlobalAccountDataEventType::Direct).await
         {
             debug!("Found direct room data in the Store, applying it");
-            self.handle_account_data(&[direct_account_data], &mut changes).await;
+            if let Ok(direct_account_data) = direct_account_data.deserialize() {
+                self.process_direct_rooms(vec![direct_account_data], &mut changes).await;
```

Comment:
And finally here, don't allocate a vector (which does an allocation on the heap, this is why it's costly), but simply pass a slice of one element.

```suggestion
                self.process_direct_rooms(&[direct_account_data], &mut changes).await;
```